JOSE_paper/paper.md

```diff
@@ -0,0 +1,254 @@
+
+In our experience working with academic developers many of the bad practices and
+anti-patterns are a result of them:
+
+- learning on the job from old code and emulating habits they observe
+- never having been made aware of the available tools and techniques
+- not keeping up to date with language/package developments
+- not investing time into writing good code, often due to placing priorities elsewhere.
+
+The intent of developing this short course, _RSE Skills (in Python)_, is to educate
+individuals in some basic software engineering principles that they can carry forward
+in their work.
+It is designed to introduce key tools and techniques to facilitate the writing of
+better code, but also to raise awareness of research software engineering and provide pointers
+to allow researchers to continue exploring and learning beyond this material.
+
+
+
+# Learning objectives
+
+The key learning objective for the workshop is to _Introduce key tools and concepts of
+research software engineering, and how they can be applied in everyday use to write
+higher-quality code_.
+
+More specifically we cover:
+
+* Why software engineering principles matter
+* (Virtual) environments and dependencies
+* Code standards and best practice (through PEP)
+  * Code formatters and linting/static-analysis tools
+* Documentation
+  * Docstrings
+  * Advice on commenting
+  * The idea of self-documenting code and variable naming
+  * READMEs and other useful files in your repository
+  * Software licenses
+* General principles for better code
+  * Magic Numbers
+  * Removal of hard-coded content to configuration files
+  * use of f-strings in Python.
+
+
+
+# Teaching materials
+
+All of the teaching materials for this course are available online in a GitHub repository.
+
+
+## Slides
+
+All of the key material is contained in a slide deck for the course
+[available online](https://jatkinson1000.github.io/rse-skills-workshop) and linked
+from the repository README.
+There is a also a [recording of the workshop](https://www.youtube.com/watch?v=nyp9t6mGdsw)
+available.
+The slides are written in [Quarto](https://quarto.org/) markdown [@Allaire_Quarto_2022]
+and rendered as [reveal.js](https://revealjs.com/) html.
+Source and instructions on how to render are included in the repository should
+others wish to tailor them to their specifications.
+
+The slides are broken into separate files by section covering:
+
+- Introduction to research software engineering, why it matters, and some examples,
+- The use of virtual environments and specifying project dependencies,
+- Code formatting for consistent readable layout,
+- Naming principles for readable, re-useable code,
+- Linting and static analysis of code,
+- Writing docstrings and comments, and
+- Other general principles including magic numbers, avoiding hard-coded values, etc.
+
+This makes it easy to remove sections to tailor the course, or for additional topics to be added in future.
```

Comment:
Should we describe this a bit more? I think I did it somewhere in a previous commit or PR, but it might have gotten lost.

Reply:
I moved this around a bit, so might be elsewhere in the paper...?
One thing I noticed when I re-read this morning was that we repeat this point a couple of times.